### PR TITLE
[batch] replace gsutil in input/output containers

### PR DESCRIPTION
**Pull request by jigold**

### CODE REVIEW

**Comment by jigold** [comment]
I don't think I quite got the parallelism option honored correctly. It's controlling the number of simultaneous files being copied but not within copies.

File: batch/batch/google_storage.py [hunk 1]
```diff
@@ -85,6 +137,21 @@ def _read_binary_gs_file(self, uri, *args, **kwargs):
...
+        file.seek(offset)
+        bucket, path = GCS._parse_uri(uri)
+        bucket = self.gcs_client.bucket(bucket)
+        f = bucket.blob(path)
+        f.metadata = {'Cache-Control': 'no-cache'}
+        f.download_to_file(file, *args, **kwargs)
+
     def _delete_gs_files(self, uri_prefix):
         bucket, prefix = GCS._parse_uri(uri_prefix)
         bucket = self.gcs_client.bucket(bucket)
```
Comment (danking) [inline batch/batch/google_storage.py]: We probably should issue these delete requests in parallel. We can benchmark and decide later though.

File: hail/python/hailtop/utils/utils.py [hunk 2]
```diff
@@ -157,8 +168,73 @@ def __init__(self, worker_pool):
...
+        if self._errors:
+            self._done.set()
+
         await self._done.wait()
 
+        if self._errors:
+            raise self._errors[0]
+
+
+class MultiWaitableSharedPool:
```
Comment (danking) [inline hail/python/hailtop/utils/utils.py]: as noted above, I don't think we need this.

File: batch/utils/copy_files.py [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+        async with CopyFileTimer(src, dest, size) as timer:
+            token = uuid.uuid4().hex[:8]
+            try:
+                starts = get_partition_starts(size, max_partitions, min_partition_size)
+                tmp_dests = [dest + f'/{token}/{uuid.uuid4().hex[:8]}' for _ in range(len(starts) - 1)]
+
+                copy_token = uuid.uuid4().hex[:5]
+                for i in range(len(starts) - 1):
+                    await copy_pool.call(copy_token, _write, timer, tmp_dests[i], starts[i], starts[i+1])
+                await copy_pool.wait(copy_token)
```
Comment (danking) [inline batch/utils/copy_files.py]: The tokens are effectively pointers to objects, we should just use objects:
```python
b = WaitableBunch(copy_pool)
for i in range(n_parts):
     await b.call(copy_token, ...)
await b.wait()
```

File: batch/utils/copy_files.py [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+    try:
+        coros = []
+        for line in sys.stdin:
+            src, dest = shlex.split(line.rstrip())
+            if '**' in src:
+                raise NotImplementedError(f'** not supported; got {src}')
+            coros.append(copies(file_pool, copy_pool, src, dest, args.max_partitions,
+                                parse_memory_in_bytes(args.min_partition_size)))
+        await asyncio.gather(*coros)
+        await file_pool.wait()
```
Comment (danking) [inline batch/utils/copy_files.py]: why do we not have to wait on the copy_pool?

File: batch/utils/copy_files.py [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+
+    file_lock_store = FileLockStore()
+
+    try:
+        coros = []
+        for line in sys.stdin:
+            src, dest = shlex.split(line.rstrip())
+            if '**' in src:
+                raise NotImplementedError(f'** not supported; got {src}')
+            coros.append(copies(file_pool, copy_pool, src, dest, args.max_partitions,
```
Comment (danking) [inline batch/utils/copy_files.py]: I'm not convinced max_partitions should be the same for reading from google storage as writing to google storage. Clearly for writing to google storage there's an externally imposed maximum.

File: batch/utils/copy_files.py [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+        return [0, 0]
+
+    if file_size / min_partition_size > max_partitions:
+        partition_size = (file_size + max_partitions - 1) // max_partitions
+    else:
+        partition_size = min_partition_size
+
+    partition_starts = [i for i in range(0, file_size, partition_size)]
+    partition_starts.append(file_size)
+    return partition_starts
```
Comment (danking) [inline batch/utils/copy_files.py]: What is the goal of this function?

I think the goal for this function is to:
- create evenly balanced partitions
given two constraints:
- the partitions can be no smaller than `min_partition_size`, and
- the number of partitions can be no more than `max_partitions`.

`get_partition_starts` does not always produce evenly balanced partitions. For example:

```
In [9]: get_partition_starts(33000, 10, 10000)                                                                                                                                                                                                                         
Out[9]: [0, 10000, 20000, 30000, 33000]
```

```
boundaries = partition_boundaries(22, 10, 7)
for x, y in zip(boundaries, boundaries[1:]):
    assert (y - x) == 7 or (y - x) == 8
```

Perfect balance is given by this many partitions: `file_size / min_partition_size`. Unfortunately, we can't have fractional partitions, so we need to redistribute the fraction evenly among all the other partitions. In general, we should be able to create a partitioning where every partition is size `x` or `x+1` for some `x` which is not much bigger or equal to the minimum partition size.

File: batch/batch/google_storage.py [hunk 7]
```diff
@@ -62,13 +99,28 @@ def _wrap_network_call(self, fun):
...
 
+    def _write_gs_file_from_filename(self, uri, filename, *args, **kwargs):
+        bucket, path = GCS._parse_uri(uri)
+        bucket = self.gcs_client.bucket(bucket)
+        f = bucket.blob(path)
+        f.metadata = {'Cache-Control': 'no-cache'}
+        f.upload_from_filename(filename, *args, **kwargs)
+
+    def _write_gs_file_from_file(self, uri, file, *args, **kwargs):
+        file.seek(0)
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch/batch/google_storage.py]: The problem is when we retry the function if it fails due to a transient error. Then the file is not in the right place.

File: batch/batch/google_storage.py [hunk 8]
```diff
@@ -85,6 +137,21 @@ def _read_binary_gs_file(self, uri, *args, **kwargs):
...
 
+    def _read_gs_file_to_filename(self, uri, filename, *args, **kwargs):
+        bucket, path = GCS._parse_uri(uri)
+        bucket = self.gcs_client.bucket(bucket)
+        f = bucket.blob(path)
+        f.metadata = {'Cache-Control': 'no-cache'}
+        f.download_to_filename(filename, *args, **kwargs)
+
+    def _read_gs_file_to_file(self, uri, file, offset, *args, **kwargs):
+        file.seek(offset)
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch/batch/google_storage.py]: See comment above about retrying the function if there's a transient error.

File: batch/batch/google_storage.py [hunk 9]
```diff
@@ -102,3 +169,41 @@ def _delete_gs_file(self, uri):
...
+        src_bucket, src_path = GCS._parse_uri(src)
+        src_bucket = self.gcs_client.bucket(src_bucket)
+        dest_bucket, dest_path = GCS._parse_uri(dest)
+        dest_bucket = self.gcs_client.bucket(dest_bucket)
+        src_f = src_bucket.blob(src_path)
+        src_bucket.copy_blob(src_f, dest_bucket, new_name=dest_path, *args, **kwargs)
+
+    def _list_gs_files(self, uri_prefix, max_results=None):
+        bucket_name, prefix = GCS._parse_uri(uri_prefix)
+        if '*' in bucket_name:
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch/batch/google_storage.py]: I do check for it in the copy_files script.

File: batch/utils/copy_files.py [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+thread_pool = None
+gcs_client = None
+copy_failure = False
+file_lock_store = None
+
+MIN_PARTITION_SIZE = 256 * 1024 * 1024
+MAX_PARTITIONS = 32
+
+
+class FileLock:
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch/utils/copy_files.py]: Yes.

File: batch/utils/copy_files.py [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+    async def __aexit__(self, exc_type, exc, tb):
+        finish_time = time.time()
+        total = finish_time - self.start_time
+        if exc_type is None:
+            print(f'copied {self.src} to {self.dest} in {total:.3f}s')
+        else:
+            print(f'failed to copy {self.src} to {self.dest} in {total:.3f}s due to {exc!r}')
+
+
+class FilePart(io.IOBase):
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch/utils/copy_files.py]: Yeah, I was going to ask what to do when everything was working and we had a close to final version.

File: batch/utils/copy_files.py (hunk 4 above)
Comment (jigold) [reply batch/utils/copy_files.py]: Because it's implicitly waited on by coroutines in the file_pool.

File: batch/utils/copy_files.py (hunk 6 above)
Comment (danking) [reply batch/utils/copy_files.py]: erm I thought about this more and maybe I am wrong. We probably want to align to some large power of two, right? On the other hand:
```
parallelism=16 block_size=2Gi replica=1
copying gs://hail-jigold-59hi5/data/large to /tmp/large of size 5.4 GB...
copied gs://hail-jigold-59hi5/data/large to /tmp/large of size 5.4 GB in 73.782s
  setup: 1.811s
  download 4294967296-5368709119: 48.524s
  download 0-2147483647: 71.478s
  download 2147483648-4294967295: 71.965s
```
This benchmark shows  the tiny partition is significantly faste.r

File: batch/batch/google_storage.py [hunk 12]
```diff
@@ -102,3 +169,41 @@ def _delete_gs_file(self, uri):
...
+
+    def _copy_gs_file(self, src, dest, *args, **kwargs):
+        src_bucket, src_path = GCS._parse_uri(src)
+        src_bucket = self.gcs_client.bucket(src_bucket)
+        dest_bucket, dest_path = GCS._parse_uri(dest)
+        dest_bucket = self.gcs_client.bucket(dest_bucket)
+        src_f = src_bucket.blob(src_path)
+        src_bucket.copy_blob(src_f, dest_bucket, new_name=dest_path, *args, **kwargs)
+
+    def _list_gs_files(self, uri_prefix, max_results=None):
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch/batch/google_storage.py:181]: max_results is just the maximum number of blobs to return overall.

File: batch/utils/copy_files.py [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+    if not os.path.exists(path):
+        raise FileNotFoundError(path)
+    if os.path.isfile(path):
+        return [(path, os.path.getsize(path))]
+    # gsutil doesn't copy empty directories
+    return flatten([listdir(path.rstrip('/') + '/' + f) for f in os.listdir(path)])
+
+
+def get_dest_path(file, src, include_recurse_dir):
+    src = src.rstrip('/').split('/')
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch/utils/copy_files.py:198]: I don't think this function is a good idea and makes things more complicated. GCS paths can have double slashes and they are distinct from single slash paths whereas unix consolidates them. I just wanted to strip the trailing slash when comparing whether the path was the same length. 

Example:
file = '/a/b/c/d.txt'
src = '/a/b/c/'

If you don't strip the trailing slash on the src, file and src will be the same length and you'll copy it to the wrong place (d.txt instead of c/d.txt).

**Comment by jigold** [comment]
I think this addresses everything except for the file partition.

File: batch/batch/google_storage.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jigold) [reply batch/batch/google_storage.py:268]: I added pagination.

File: batch/batch/google_storage.py [hunk 14]
```diff
@@ -102,3 +169,41 @@ def _delete_gs_file(self, uri):
...
+        if '*' in bucket_name:
+            bucket_prefix = bucket_name.split('*')[0]
+            buckets = [bucket for bucket in self.gcs_client.list_buckets(prefix=bucket_prefix, max_results=max_results)
+                       if fnmatch.fnmatch(bucket.path.replace('/b/', 'gs://'), bucket_name)]
+        else:
+            buckets = [self.gcs_client.bucket(bucket_name)]
+
+        for bucket in buckets:
+            for blob in bucket.list_blobs(prefix=prefix):
+                yield (urllib.parse.unquote(blob.public_url.replace('https://storage.googleapis.com/', 'gs://')), blob.size)
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch/batch/google_storage.py]: Didn't see that `name` was an option in the docs but it makes sense.

File: batch/test/test_copy_files.py [hunk 15]
```diff
@@ -0,0 +1,229 @@
+import os
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch/test/test_copy_files.py:1]: Added a test for multiple wildcards.

File: batch/utils/copy_files.py (hunk 4 above)
Comment (jigold) [reply batch/utils/copy_files.py]: I added a wait for it now that I restructured the waitable pool

File: batch/utils/copy_files.py (hunk 6 above)
Comment (jigold) [reply batch/utils/copy_files.py]: I made the partitions as balanced as possible

File: batch/utils/copy_files.py (hunk 5 above)
Comment (jigold) [reply batch/utils/copy_files.py]: I changed this so reading doesn't have max_partitions but upload does and the max is 32 by default.

File: batch/batch/google_storage.py (hunk 1 above)
Comment (jigold) [reply batch/batch/google_storage.py:244]: I'm not going to worry about this right now. I can add an asana task.

**Comment by danking** [comment]
Cotton is on the job.

**Comment by jigold** [comment]
This now has the multiprocessing in it for just reading and writing. The rest of the GCS functions go through the thread pool. I think we should eventually fix the other read/write operations that currently go through the thread pool to go in a process pool, but that can be a different PR unless you see an easy way to do it.

**Comment by jigold** [comment]
This is all passing except check_batch is running in an image on python 3.6 and the worker is on python 3.7. The process initializer isn't exposed until 3.7. I think I need to change the base image, but am worried about breaking Hail and other dependencies. Otherwise, I can just pass the key file path and the project and create the credentials and gcs client each time a function is called rather than once per process. This will make it slower for small files.

**Comment by jigold** [comment]
Stacked on #8440

File: batch/batch/google_storage.py [hunk 16]
```diff
@@ -102,3 +136,41 @@ def _delete_gs_file(self, uri):
...
+
+    def _copy_gs_file(self, src, dest, *args, **kwargs):
+        src_bucket, src_path = GCS._parse_uri(src)
+        src_bucket = self.gcs_client.bucket(src_bucket)
+        dest_bucket, dest_path = GCS._parse_uri(dest)
+        dest_bucket = self.gcs_client.bucket(dest_bucket)
+        src_f = src_bucket.blob(src_path)
+        src_bucket.copy_blob(src_f, dest_bucket, new_name=dest_path, *args, **kwargs)
+
+    def _list_gs_files(self, uri_prefix, max_results=None):
```
Comment (danking) [inline batch/batch/google_storage.py:148]: Fun fact: `gsutil ls` does not support stars in the bucket name!

Comment (danking) [inline batch/batch/google_storage.py:148]: Fun fact: gsutil ls does not support stars in the bucket name! Should we?

File: batch/batch/google_storage.py [hunk 17]
```diff
@@ -102,3 +136,41 @@ def _delete_gs_file(self, uri):
...
+        bucket_name, prefix = GCS._parse_uri(uri_prefix)
```
Comment (danking) [inline batch/batch/google_storage.py:149]: Can we add:
```
assert '**' not in bucket_name
assert '?' not in bucket_name
```
I'm worried we'll forget and use this method elsewhere.

File: batch/test/test_dag.py [hunk 18]
```diff
@@ -2,6 +2,7 @@
 import time
 import re
 import pytest
+import uuid
```
Comment (danking) [inline batch/test/test_dag.py:5]: unnecessary change.

File: ci/ci/ci.py [hunk 19]
```diff
@@ -261,10 +261,11 @@
     batch_client = request.app['batch_client']
 
     async def get_failure_information(batch):
+        jobs = await collect_agen(batch.jobs())
         return [
             {**j,
              'log': await batch_client.get_job_log(j['batch_id'], j['job_id'])}
-            for j in batch.jobs() if j['state'] != 'Success']
+            for j in jobs if j['state'] != 'Success']
```
Comment (danking) [inline ci/ci/ci.py:268]: you moved this to another PR, can we remove it from here?

File: hail/python/hailtop/utils/utils.py [hunk 20]
```diff
@@ -44,7 +45,7 @@ def async_to_blocking(coro):
 
 async def blocking_to_async(thread_pool, fun, *args, **kwargs):
     return await asyncio.get_event_loop().run_in_executor(
-        thread_pool, lambda: fun(*args, **kwargs))
+        thread_pool, functools.partial(fun, *args, **kwargs))
```
Comment (danking) [inline hail/python/hailtop/utils/utils.py:48]: you don't like my thunks? 😉

File: hail/python/hailtop/utils/utils.py [hunk 21]
```diff
@@ -158,8 +170,54 @@ def __init__(self, worker_pool):
...
+
+        await self.shared_pool._worker_pool.call(invoke)
+
+    async def wait(self):
+        assert not self._waiting
+        self._waiting = True
+        if self._n_complete == self._n_submitted:
+            self._done.set()
+        if self._errors:
+            self._done.set()
```
Comment (danking) [inline hail/python/hailtop/utils/utils.py:214]: It seems simpler to remove `_waiting`, and just always set self._done when we reach all tasks complete, then here we needn't have either check. If it already happened, then done will be set. Otherwise, it will be set while we're waiting on it.

File: hail/python/hailtop/utils/utils.py [hunk 22]
```diff
@@ -158,8 +170,54 @@ def __init__(self, worker_pool):
...
+
+        await self._done.wait()
+
+        if self._errors:
+            raise self._errors[0]
```
Comment (danking) [inline hail/python/hailtop/utils/utils.py:219]: I'm inclined to raise a new error that at least includes all the names of the errors, in the odd case that there are many different kinds. Still good to `... from self._errors[0]`

File: batch/batch/google_storage.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (danking) [reply batch/batch/google_storage.py:148]: Erm I should say, it *acts on* glob stars in the bucket name, but nowhere else.

File: batch/batch/worker/copy_files.py [hunk 23]
```diff
@@ -0,0 +1,380 @@
...
+def listdir(path):
+    if not os.path.exists(path):
+        raise FileNotFoundError(path)
+    if os.path.isfile(path):
+        return [(path, os.path.getsize(path))]
+    # gsutil doesn't copy empty directories
+    return flatten([listdir(path + '/' + f) for f in os.listdir(path)])
+
+
+def get_dest_path(file, src, include_recurse_dir):
```
[the remark this answers is not in the document]
Comment (danking) [reply batch/batch/worker/copy_files.py:165]: Eh, maybe not. I'm not sure, now that I understand the pervasive use of the term `src`, that my comment helps.

File: batch/batch/worker/copy_files.py [hunk 24]
```diff
@@ -0,0 +1,380 @@
...
+                         fnmatch.fnmatchcase(path, src.rstrip('/'))]
+    else:
+        src = os.path.abspath(src)
+        src_paths = glob.glob(src, recursive=True)
+        src_paths = flatten([listdir(src_path) for src_path in src_paths])
+
+    if len(src_paths) == 1:
+        file, size = src_paths[0]
+        if dest.endswith('/'):
+            paths = [(file, f'{dest}{os.path.basename(file)}', size)]
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch/batch/worker/copy_files.py:309]: No, it ignores multiple slashes.

```
>>> os.listdir('/tmp/a/b/c/')
['d.txt']
>>> os.listdir('/tmp/a/b//c/')
['d.txt']
>>> os.listdir('/tmp//a/b//c/')
['d.txt']
>>> os.listdir('//tmp//a/b//c/')
['d.txt']
```

File: hail/python/hailtop/utils/utils.py (hunk 20 above)
Comment (jigold) [reply hail/python/hailtop/utils/utils.py:48]: The process pool won't work with them. This is what you have to do to get kwargs

File: batch/batch/worker/copy_files.py [hunk 25]
```diff
@@ -0,0 +1,380 @@
...
+        src_prefix = re.split('\\*|\\[\\?', src)[0].rstrip('/')
+        maybe_src_paths = [('gs://' + blob.bucket.name + '/' + blob.name, blob.size)
+                           for blob in await batch_gcs.list_gs_files(src_prefix)]
+        non_recursive_matches = [(path, size) for path, size in maybe_src_paths if fnmatch.fnmatchcase(path, src)]
+        if not src.endswith('/') and non_recursive_matches:
+            src_paths = non_recursive_matches
+        else:
+            src_paths = [(path, size) for path, size in maybe_src_paths
+                         if fnmatch.fnmatchcase(path, src.rstrip('/') + '/*') or
+                         fnmatch.fnmatchcase(path, src.rstrip('/'))]
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch/batch/worker/copy_files.py:300]: This code doesn't work. We're not matching for an exact path. I don't see what the ? buys us. It only matches a single character.

File: hail/python/hailtop/utils/utils.py (hunk 21 above)
Comment (jigold) [reply hail/python/hailtop/utils/utils.py:214]: I don't think you can do this.

File: hail/python/hailtop/utils/utils.py (hunk 22 above)
Comment (jigold) [reply hail/python/hailtop/utils/utils.py:219]: I don't think this is necessary.

File: batch/test/test_copy_files.py [hunk 26]
```diff
@@ -0,0 +1,264 @@
...
+import re
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch/test/test_copy_files.py:2]: > Note: If you use the Google Cloud Platform Console to create folders, it does so by creating a "placeholder" object that ends with a "/" character. gsutil skips these objects when downloading from the cloud to the local file system, because attempting to create a file that ends with a "/" is not allowed on Linux and macOS. Because of this, it is recommended that you not create objects that end with "/" (unless you don't need to be able to download such objects using gsutil).

File: hail/python/hailtop/utils/utils.py (hunk 20 above)
Comment (danking) [reply hail/python/hailtop/utils/utils.py:52]: Heh. Unfortunate that lambdas are not pickle-able by default.

File: hail/python/hailtop/utils/utils.py (hunk 21 above)
Comment (danking) [reply hail/python/hailtop/utils/utils.py:218]: Can you explain what I misunderstand? It seems like the timeline of an object like this is:
1. get created
2. add some jobs via `call`
3. `wait` once on this bunch

If that understanding is true, then `done` starts out unset and only transitions to `set` if the bunch is finished (due either to an error or to all jobs finishing). If we unconditionally set it when there's an error or when all jobs finish, then when we enter wait either it is set or it will be set in the future (assuming the jobs terminate). There's no need to check any conditions in `wait`.

File: build.yaml [hunk 27]
```diff
@@ -1120,7 +1121,35 @@ steps:
...
+    - batch_worker_image
     - deploy_batch
+ - kind: runImage
+   name: test_batch_copy_files
+   image:
+     valueFrom: batch_image.image
+   script: |
+     export HAIL_DEPLOY_CONFIG_FILE=/deploy-config/deploy-config.json
+     export PROJECT={{ global.project }}
+     # FIXME parallelize this
```
Comment (danking) [inline build.yaml:1133]: reminder to parallelize this like test_hail_python

File: build.yaml [hunk 28]
```diff
@@ -1120,7 +1121,35 @@ steps:
...
+      namespace:
+        valueFrom: default_ns.name
+      mountPath: /deploy-config
+   dependsOn:
+    - default_ns
+    - batch_pods_ns
+    - copy_files
+    - base_image
+    - service_base_image
+    - batch_image
```
Comment (danking) [inline build.yaml:1156]: We just need `batch_image` and  `copy_files`, we don't reference any of the other things

File: batch/test/test_copy_files.py [hunk 29]
```diff
@@ -0,0 +1,547 @@
...
+def upload_files(src, dest):
+    os.system(f'gsutil -m -q cp -r {src} {dest}')
+
+
+def move_files(src, dest):
+    os.system(f'gsutil -m -q mv -r {src} {dest}')
+
+
+def remove_remote_dir(path):
+    os.system(f'gsutil -m -q rm -r {path}')
```
Comment (danking) [inline batch/test/test_copy_files.py:45]: All of these paths should be shell quoted if we're executing them in a shell.

Actually, we should just use one of the subprocess functions to avoid shell escaping entirely.

File: batch/test/test_copy_files.py [hunk 30]
```diff
@@ -0,0 +1,547 @@
...
+  -f {shq(src)} {shq(dest)} 2>&1
+'''
+    result = sp.run(cmd, shell=True, stdout=sp.PIPE, stderr=sp.PIPE)
+    return str(result.stdout), result.returncode
+
+
+def cp_gsutil(src, dest):
+    cmd = f'gcloud -q auth activate-service-account --key-file={key_file}; gsutil -m -q cp -r {shq(src)} {shq(dest)} 2>&1'
+    result = sp.run(cmd, shell=True, stdout=sp.PIPE, stderr=sp.PIPE)
+    return str(result.stdout), result.returncode
```
Comment (danking) [inline batch/test/test_copy_files.py:77]: I think it should be sufficient to activate the service account once in the build.yaml script and then all subsequent calls will read the credentials from the file system.

File: batch/test/test_copy_files.py [hunk 31]
```diff
@@ -0,0 +1,547 @@
...
+  --key-file {key_file} \\
+  --project {project} \\
+  --parallelism {parallelism} \\
+  --min-partition-size {min_partition_size} \\
+  --max-upload-partitions {max_upload_partitions} \\
+  --max-download-partitions {max_download_partitions} \\
+  -f {shq(src)} {shq(dest)} 2>&1
+'''
+    result = sp.run(cmd, shell=True, stdout=sp.PIPE, stderr=sp.PIPE)
+    return str(result.stdout), result.returncode
```
Comment (danking) [inline batch/test/test_copy_files.py:71]: Yeah, you're already using subprocess here, I'd avoid the shell entirely so you don't have any quoting nonsense:

```
result = sp.run(['python3', '-u', '-m', 'batch.worker.copy_files',
                 '--key-file', key_file,
....
                 '-f', src, dest], stdout=sp.PIPE, stderr=sp.STDOUT)
```

File: batch/test/test_copy_files.py [hunk 32]
```diff
@@ -0,0 +1,547 @@
...
+        assert 'IsADirectoryError' in _get_output(result, 'll', 'batch'), _get_output(result, 'll', 'batch')
+
+    def test_download_file_invalid_dest_dir_with_wildcard(self):
+        result = self.assert_batch_same_as_gsutil('/data/*', '/b/')
+
+        assert 'destination must name a directory when matching multiple files' in _get_output(result, 'rl', 'batch'), _get_output(result, 'rl', 'batch')
+        assert 'NotADirectoryError' in _get_output(result, 'rl', 'batch'), _get_output(result, 'rl', 'batch')
+
+        assert 'destination must name a directory when matching multiple files' in _get_output(result, 'll', 'batch'), _get_output(result, 'll', 'batch')
+        assert 'NotADirectoryError' in _get_output(result, 'll', 'batch'), _get_output(result, 'll', 'batch')
```
Comment (danking) [inline batch/test/test_copy_files.py:366]: The last there methods don't assert anything about rr or lr, is that intentional?

File: batch/test/test_copy_files.py [hunk 33]
```diff
@@ -0,0 +1,547 @@
+import os
+import unittest
```
Comment (danking) [inline batch/test/test_copy_files.py:2]: do we have a test for downloading files when there exists a size zero file named `foo/` and also some files named `foo/a` and `foo/b`? Since Hail tables and matrix tables have this structure, I think it's important we test that behavior.

File: batch/build-batch-worker-image-startup.sh [hunk 34]
```diff
@@ -30,6 +30,5 @@ curl -fsSL "https://github.com/GoogleCloudPlatform/docker-credential-gcr/release
 docker-credential-gcr configure-docker
 
 docker pull ubuntu:18.04
```
Comment (danking) [inline batch/build-batch-worker-image-startup.sh:32]: why do we pull this?

File: batch/batch/worker/copy_files.py [hunk 35]
```diff
@@ -0,0 +1,341 @@
...
+        if len(glob_paths) == 1:
+            return [(file, dest, size) for file, size in glob_paths]
+        raise NotADirectoryError
+    else:
+        assert is_dir_src and is_dir_dest
+        if len(glob_paths) == 0:
+            raise FileNotFoundError
+
+        # https://cloud.google.com/storage/docs/gsutil/commands/cp#how-names-are-constructed_1
+        include_recurse_dir = not is_gcs_path(dest) or await is_gcs_dir(dest)  # DOUBLE CHECK THIS!!!
```
Comment (danking) [inline batch/batch/worker/copy_files.py]: double check this

File: hail/python/hailtop/utils/os.py [hunk 36]
```diff
@@ -0,0 +1,153 @@
...
+        elif c[i] in wildcards:
+            return ''.join(new_c)
+        new_c.append(c[i])
+        i += 1
+    return ''.join(new_c)
+
+
+def makedirs(path):
+    path = os.path.abspath(path)
+    os.makedirs(os.path.dirname(path), exist_ok=True)
```
[the remark this answers is not in the document]
Comment (jigold) [reply hail/python/hailtop/utils/os.py:75]: The reason for the abs path was in case someone gave a relative path as input to the function.

File: hail/python/hailtop/utils/os.py [hunk 37]
```diff
@@ -0,0 +1,153 @@
...
+    with open(path, 'ab') as fp:
+        fp.truncate(size)
+
+
+def _glob(path):
+    is_dir = path.endswith('/')
+    path = os.path.abspath(path)
+    if is_dir:
+        path += '/'
+    paths = glob.glob(escape(path), recursive=True)
```
[the remark this answers is not in the document]
Comment (jigold) [reply hail/python/hailtop/utils/os.py:105]: I want the absolute paths to be returned. I refactored this a bit and is better now.

File: hail/python/hailtop/utils/utils.py [hunk 38]
```diff
@@ -158,8 +174,54 @@ async def wait(self):
         self._waiting = True
         if self._n_complete == self._n_submitted:
             self._done.set()
+        if not self._ignore_errors and self._errors:
+            self._done.set()
```
[the remark this answers is not in the document]
Comment (jigold) [reply hail/python/hailtop/utils/utils.py:178]: The issue is if we remove the self._waiting, then we add the first function to the queue and that runs successfully before we get a chance to add the second function to the queue. At that point, the n_complete will equal the n_submitted and we'll think we're done and done will be set. Then say the second function is long running and we call wait. It will instantly return because the first function set done.

File: hail/python/hailtop/utils/utils.py (hunk 21 above)
Comment (jigold) [reply hail/python/hailtop/utils/utils.py:218]: See the above comment.

File: batch/test/test_copy_files.py (hunk 32 above)
Comment (jigold) [reply batch/test/test_copy_files.py:366]: We check everything is the same in `self.assert_batch_same_as_gsutil`. The point of the additional asserts is to make sure I actually printed the error message I intended and not a different error message.

File: batch/test/test_copy_files.py (hunk 33 above)
Comment (jigold) [reply batch/test/test_copy_files.py:2]: It's impossible to create a size zero file named `foo/` just with gsutil alone. Let me see if I can do it with the python batch client.

File: batch/build-batch-worker-image-startup.sh (hunk 34 above)
Comment (jigold) [reply batch/build-batch-worker-image-startup.sh:32]: We want to cache ubuntu 18:04

File: batch/batch/worker/copy_files.py [hunk 39]
```diff
@@ -0,0 +1,342 @@
...
+                    return []
+                print('destination is a directory')
+                raise IsADirectoryError
+            if len(glob_paths) > 1:
+                print('destination must name a directory when matching multiple files')
+                raise NotADirectoryError
+
+    is_dir_src = await is_dir(src)
+    is_dir_dest = await is_dir(dest)
+
```
Comment (danking) [inline batch/batch/worker/copy_files.py:164]:
```
if len(glob_paths) == 0:
    raise FileNotFoundError
```

Appears in every case, let's lift it to the top of this method, before the first `if`.

File: batch/batch/worker/copy_files.py [hunk 40]
```diff
@@ -0,0 +1,342 @@
...
+    if not is_dir_src and not is_dir_dest:
+        if len(glob_paths) == 0:
+            raise FileNotFoundError
+        if len(glob_paths) == 1:
+            return [(file, dest, size) for file, size in glob_paths]
+        if is_gcs_path(dest):
+            include_recurse_dir = not is_gcs_path(dest) or await is_gcs_dir(dest)  # DOUBLE CHECK THIS!!!
```
Comment (danking) [inline batch/batch/worker/copy_files.py:171]: `is_gcs_path(dest)` must be `True` and `is_gcs_dir(dest)` must be `False`, so we can just set this to `False`

File: batch/batch/worker/copy_files.py (hunk 39 above)
Comment (jigold) [reply batch/batch/worker/copy_files.py:164]: I don't think this is correct. There's a bug here. We should be able to copy an empty local directory with no errors.
